js/debugger.ts

```diff
@@ -0,0 +1,59 @@
+import { debug, toHex } from './util';
+import { word } from './types';
+
+import SYMBOLS from './symbols';
+import CPU6502, { DebugInfo } from './cpu6502';
+
+const alwaysBreak = (_info: DebugInfo) => { return true; };
+
+export default class Debugger {
+    private TRACE = false;
+    private MAX_TRACE = 256;
+    private trace: DebugInfo[] = [];
+    private breakpoints: Map<number, (info: DebugInfo) => boolean> = new Map()
+
+    constructor(private cpu: CPU6502) {}
+
+    stepCycles(step: number) {
+        this.cpu.stepCyclesDebug(this.TRACE ? 1 : step, () => {
+            const info = this.cpu.getDebugInfo();
+            const addr = info[0];
+
+            if (this.breakpoints.get(addr)?.(info)) {
+                debug(this.cpu.printDebugInfo(info, SYMBOLS));
+                stop();
```

Comment:
I don't see this defined. What am I missing?

Reply:
Sigh. `stop()` is a browser global so typescript didn't save me there. I need to tie this to the `run()` in apple2.ts